vscode/src/projectTemplates.ts

```diff
@@ -0,0 +1,329 @@
+            } catch (Err) {
+                // cannot decipher text from encoding, assume raw data
+                return data;
+            }
+        } else {
+            str = data;
+        }
+        while (match = regex.exec(str)) {
+            let key = match[1];
+            let val: string | undefined = placeholders[key];
+            if (!val) {
+                let variableInput = <vscode.InputBoxOptions>{
+                    prompt: `Please enter a value for: ` + match[0]
+                };
+                val = await window.showInputBox(variableInput).then(
+                    value => {
+                        if (value) {
+                            // update map
+                            placeholders[key] = value;
+                        }
+                        return value;
+                    }
+                );
+            }
+            ++nmatches;
+        }
+        // reset regex
+        regex.lastIndex = 0;
+        // compute output
+        let out: string | Buffer = data;
+        if (nmatches > 0) {
+            // replace placeholders in string
+            str = str.replace(regex,
+                (match, key) => {
+                    let val = placeholders[key];
+                    if (!val) {
+                        val = match;
+                    }
+                    return val;
+                }
+            );
+            // if input was a buffer, re-encode to buffer
+            if (Buffer.isBuffer(data)) {
+                out = Buffer.from(str, encoding);
+            } else {
+                out = str;
+            }
+        }
+        return out;
+    }
+
+    /**
+     * Populates a workspace folder with the contents of a template.
+     * @param workspace Current workspace folder to populate.
+     * @param template Relative path to the template folder.
+     * @param placeholderValues Placeholder key-value pairs.
+     */
+    public async createFromTemplate(workspace: string, template: string, placeholderValues: any) {
+
+        // get template folder
+        let templateRoot = await this.getTemplatesDir();
+        let templateDir = path.join(templateRoot, template);
+        if (!fs.existsSync(templateDir) || !fs.lstatSync(templateDir).isDirectory()) {
+            window.showErrorMessage("Template '" + data.find(x => x.id == template).name + "' does not exist.");
+            return undefined;
+        }
+        // update placeholder configuration
+        let usePlaceholders = this.config.get("usePlaceholders", false);
+        let placeholderRegExp = this.config.get("placeholderRegExp", "\\${(\\w+)?}");
+        // let placeholders: { [placeholder: string]: string | undefined } = this.config.get("placeholders", {});
+        let placeholders = placeholderValues;
+        // re-read configuration, merge with current list of placeholders
+        let newplaceholders: { [placeholder: string]: string } = this.config.get("placeholders", {});
```

Comment:
newplaceholders => Shall we use camel case notation?

Reply:
Done.